### PR TITLE
Implement BlockItemData.

### PR DESCRIPTION
**Pull request by kenzierocks**
Implementation of `BlockItemData`.
- [X] Registration of field getters and setters
- [X] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableBlockItemData`
- [X] Accurately extended the appropriate `AbstractImmutableData` implementation
- [X] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [X] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `BlockItemDataProcessor`(s):
- [X] Appropriately extended `AbstractSpongeValueProcessor`
- [X] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [X] Registered the `Key` correctly by `item_blockstate` in the `KeyRegistry`
- [X] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [X] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code: [100% *AWESOME* since 11/9](https://github.com/kenzierocks/SpongePlugins/blob/master/src/main/java/me/kenzierocks/plugins/TestSpongeBlockItemData.java)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeBlockItemData.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.spongepowered.common.util.BlockUtil;
+
+public class ImmutableSpongeBlockItemData extends AbstractImmutableSingleData<BlockState, ImmutableBlockItemData, BlockItemData>
+        implements ImmutableBlockItemData {
+
+    private final ImmutableValue<BlockState> state;
+
+    public ImmutableSpongeBlockItemData(BlockState value) {
+        super(ImmutableBlockItemData.class, value, Keys.ITEM_BLOCKSTATE);
+        this.state = new ImmutableSpongeValue<>(Keys.ITEM_BLOCKSTATE, BlockTypes.STONE.getDefaultState(), value);
```
Comment (lucdon) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeBlockItemData.java]: BlockTypes.STONE.getDefaultState() should be a dataConstant so move this to `DataConstants`

Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeBlockItemData.java]: @lucdon Did you not see the comment in the description?
> N.B.: @clienthax is supposed to be adding a DataConstant for the BlockTypes.STONE.getDefaultState() calls. I'll update them when his PR is merged, or I can add it as well.

Comment (lucdon) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeBlockItemData.java]: oops did not see that 1

**Comment by kenzierocks** [comment]
Considering Mr. Hax's PR adds a `FALLING_BLOCK` specific `DataConstant` that actually has slightly more meaning for only falling blocks, I'll be adding my own `DataConstant` for the `BLOCK_ITEM` default state.

File: src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java [hunk 2]
```diff
@@ -967,9 +973,9 @@ public static void setupSerialization(Game game) {
         dataRegistry.registerDataProcessorAndImpl(TargetLivingData.class, SpongeTargetLivingData.class, ImmutableTargetLivingData.class,
                 ImmutableSpongeTargetLivingData.class, targetLivingDataProcessor);
 
-        final FireworkEffectDataProcessor fireworkEffectDataProcessor = new FireworkEffectDataProcessor();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java]: Why are you removing this?

Comment (kenzierocks) [inline src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java]: Merge fail.

Comment (kenzierocks) [inline src/main/java/org/spongepowered/common/data/SpongeSerializationRegistry.java]: Fixed now.

**Pull request closed without merging**